src/main/host/syscall/unistd.rs

```diff
@@ -107,18 +144,34 @@ fn read_helper(
     let desc = match desc {
         CompatDescriptor::New(d) => d,
         CompatDescriptor::Legacy(_) => unsafe {
-            return c::syscallhandler_read(
+            return c::syscallhandler_pread64(
                 sys as *mut c::SysCallHandler,
                 args as *const c::SysCallArgs,
             );
         },
     };
 
+    read_helper(sys, fd, desc, buf_ptr, buf_size, offset)
+}
+
+fn read_helper(
+    sys: &mut c::SysCallHandler,
+    fd: libc::c_int,
+    desc: &Descriptor,
+    buf_ptr: c::PluginPtr,
+    buf_size: libc::size_t,
+    offset: libc::off_t,
+) -> c::SysCallReturn {
     // need a non-null buffer
     if buf_ptr.val == 0 {
         return SyscallReturn::Error(nix::errno::Errno::EFAULT).into();
     }
 
+    // we can only seek on files, otherwise its a pipe error
```

Comment:
Got a bit confused here until I realized that (at least currently) this function is only called for pipe descriptors. If that'll always be the case, then maybe `desc` should specifically be a pipe descriptor type? If it's not, then maybe some code validating that it's a pipe descriptor (for now)?

Reply:
I modified this a bit to pass the offset on to the file object (for example the `PipeFile::read()`). Then since pipes don't support seeking, it returns the ESPIPE error from there). Hopefully it makes more sense now.